### PR TITLE
Support SPLUNK_REALM and SPLUNK_ACCESS_TOKEN

### PR DESCRIPTION
**Pull request by RassK**
### What

Addresses:
* https://signalfuse.atlassian.net/browse/APMI-3235
* https://signalfuse.atlassian.net/browse/APMI-3234

### Next

Support ASP.NET web.config configuration
https://github.com/signalfx/splunk-otel-dotnet/pull/78/files#diff-e1c950e36ab40bff7a0777a1e81fabd11c5c255b925b203c356656058e4880eeR51

### CODE REVIEW

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Configuration/CompositeConfigurationSource.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
+﻿// <copyright file="CompositeConfigurationSource.cs" company="Splunk Inc.">
```
Comment (pellared) [inline src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Configuration/CompositeConfigurationSource.cs:1]: It is from OTel. We need to keep the original license header. See how it is done e.g. in https://github.com/signalfx/splunk-otel-dotnet/blob/main/test/Splunk.OpenTelemetry.AutoInstrumentation.IntegrationTests/SmokeTests.cs

Same comment applies to other places

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/AssemblyInfo.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Splunk.OpenTelemetry.AutoInstrumentation.Plugin.Tests")]
```
Comment (Kielek) [inline src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/AssemblyInfo.cs:19]: I think that you can remove whole file https://github.com/signalfx/splunk-otel-dotnet/blob/0006c29c5f6a99ef4e26c2ca2656a89325f3fa4a/src/Directory.Build.props#L36-L38

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Configuration/CompositeConfigurationSource.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Splunk.OpenTelemetry.AutoInstrumentation.Plugin.Configuration;
```
Comment (Kielek) [inline src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Configuration/CompositeConfigurationSource.cs:22]: Looks like a copy from upstream repo. If so, we need add some headers here.

It is applicable for most of the newly added files.

@dszmigielski could you please advice?

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs [hunk 4]
```diff
@@ -14,22 +14,71 @@
...
+        /// <summary>
+        /// Configure Otlp exporter options
+        /// </summary>
+        /// <param name="options">Otlp options</param>
+        public void ConfigureOptions(OtlpExporterOptions options)
+        {
+            if (_settings.Realm != null)
+            {
+                options.Endpoint = new Uri(string.Format(Constants.Ingest.TracesIngestTemplate, _settings.Realm));
+            }
```
Comment (pellared) [inline src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs:67]: I think this breaks the following requirement:

> If both SPLUNK_REALM and OTEL_EXPORTER_*_ENDPOINT are set then OTEL_EXPORTER_*_ENDPOINT takes precedence.

From: https://github.com/signalfx/gdi-specification/blob/main/specification/configuration.md

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs [hunk 5]
```diff
@@ -14,22 +14,71 @@
...
+
+        /// <summary>
+        /// Configures Traces
+        /// </summary>
+        /// <param name="builder"><see cref="TracerProviderBuilder"/>  to configure</param>
+        /// <returns>Returns <see cref="TracerProviderBuilder"/> for chaining.</returns>
+        public TracerProviderBuilder ConfigureTracerProvider(TracerProviderBuilder builder)
+        {
+            return builder;
+        }
```
Comment (pellared) [inline src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs:56]: This is not needed. Right?

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Configuration/EnvironmentConfigurationSource.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+
+namespace Splunk.OpenTelemetry.AutoInstrumentation.Plugin.Configuration
```
Comment (Kielek) [inline src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Configuration/EnvironmentConfigurationSource.cs:19]: Can you apply filescoped namespaced? Whole PR.

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs [hunk 7]
```diff
@@ -14,22 +14,71 @@
...
+
+        /// <summary>
+        /// Configure Otlp exporter options
+        /// </summary>
+        /// <param name="options">Otlp options</param>
+        public void ConfigureOptions(OtlpExporterOptions options)
```
Comment (pellared) [inline src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs:62]: We would need to document that direct ingest is supported only when OTLP exporter is used.
Alternatively, we can add support for other exporters, but I think it is YAGNI.

**Review by Kielek — COMMENTED** [review]
Probably not related to this PR. But we need to update documentation how to enable this plugin (by default)

**Review by dszmigielski — APPROVED** [review]

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Configuration/CompositeConfigurationSource.cs (hunk 3 above)
Comment (dszmigielski) [reply src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Configuration/CompositeConfigurationSource.cs:22]: If its a copy from otel then we need to add both our header as well as otel one.
Example:
https://github.com/signalfx/splunk-otel-dotnet/blob/main/test/Splunk.OpenTelemetry.AutoInstrumentation.IntegrationTests/Helpers/EnvironmentHelper.cs

**Review by pellared — COMMENTED** [review]
Looks nice so far. Have you tested it E2E with the latest OTel .NET AutoInstr release and our backend?

Missing things:
1. Documentation
2. Scripts should inject the plugin
3. Probably we need to add the plugin to the archive in the build pipeline

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs (hunk 4 above)
Comment (pellared) [reply src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs:67]: We should also add a unit test for it

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs [hunk 8]
```diff
@@ -14,22 +14,71 @@
...
+        /// <param name="settings">PluginSettins instance</param>
+        internal Traces(PluginSettings settings)
+        {
+            _settings = settings;
+        }
+
+        /// <summary>
+        /// Configures Traces
+        /// </summary>
+        /// <param name="builder"><see cref="TracerProviderBuilder"/>  to configure</param>
```
Comment (pellared) [inline src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs:51]:
```suggestion
        /// <param name="builder"><see cref="TracerProviderBuilder"/> to configure</param>
```

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Metrics.cs [hunk 9]
```diff
@@ -14,22 +14,71 @@
...
+        /// <param name="settings">PluginSettings instance</param>
+        internal Metrics(PluginSettings settings)
+        {
+            _settings = settings;
+        }
+
+        /// <summary>
+        /// Configures Metrics
+        /// </summary>
+        /// <param name="builder"><see cref="MeterProviderBuilder"/>  to configure</param>
```
Comment (pellared) [inline src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Metrics.cs:51]:
```suggestion
        /// <param name="builder"><see cref="MeterProviderBuilder"/> to configure</param>
```

File: src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs (hunk 5 above)
Comment (Kielek) [reply src/Splunk.OpenTelemetry.AutoInstrumentation.Plugin/Traces.cs:56]: Lets keep it. I will extend in following PR.

**Comment by RassK** [comment]
> Missing things:
> 
> 1. Documentation
> 2. Scripts should inject the plugin
> 3. Probably we need to add the plugin to the zip archive in the build pipeline

Current state:
1-2 should be still fixed, 3 is already done by @Kielek (See `nuke PackSplunkDistribution`)

**Review by Kielek — APPROVED** [review]

**Comment by RassK** [comment]
> > Missing things:
> > 
> > 1. Documentation
> > 2. Scripts should inject the plugin
> > 3. Probably we need to add the plugin to the zip archive in the build pipeline
> 
> Current state:
> 1-2 should be still fixed, 3 is already done by @Kielek (See `nuke PackSplunkDistribution`)

3. Actually we are packing the plugin to the wrong directory. It cannot be in the separate dir.

**Comment by dszmigielski** [comment]
> 

https://github.com/signalfx/splunk-otel-dotnet/pull/82

**Comment by Kielek** [comment]
Merging, lets handle documentation and scripts in separate PR.

**Pull request merged**